src/main/java/fi/aalto/cs/apluscourses/dal/TokenAuthentication.java

```diff
@@ -20,8 +19,8 @@
    * @param token           Token.  The array is cloned and the argument is not cleared.
    * @param passwordStorage Password storage.  If null, the authentication won't be persisted.
    */
-  public TokenAuthentication(@NotNull String user,
-                             @NotNull char[] token,
+  protected TokenAuthentication(@NotNull String user,
+                                char[] token,
```

Comment:
Just curious: What's wrong with `@NotNull` arrays?

Reply:
`Primitive type members cannot be annotated`

```java
org.jetbrains.annotations @Documented 
@Retention(RetentionPolicy.CLASS) 
@Target({ElementType.METHOD,ElementType.FIELD,ElementType.PARAMETER,ElementType.LOCAL_VARIABLE,ElementType.TYPE_USE}) 
public interface NotNull
extends annotation.Annotation
```

`An element annotated with NotNull claims null value is forbidden to return (for methods), pass to (parameters) and hold (local variables and fields). Apart from documentation purposes this annotation is intended to be used by static analysis tools to validate against probable runtime errors and element contract violations.
`

Reply:
But is `char[]` a primitive?

Reply:
char would be, but arrays not

Reply:
`char @NotNull [ ] array` 😉 
thanks to @nikke234